komga/src/main/kotlin/org/gotson/komga/domain/service/LibraryScanner.kt

```diff
@@ -18,95 +26,238 @@ class LibraryScanner(
-
-            // sort all books
-            seriesLifecycle.sortBooks(existingSeries)
+  private fun restoreDeletedBooks(existingSeries: Series, existingBooks: Collection<Book>, booksToCheck: Collection<Book>): Collection<Book> {
+    val booksToRestore = booksToCheck.asSequence()
+      .filterNot { newBook -> existingBooks.map { it.fileHash }.contains(newBook.fileHash) }
+      .map { it to matchExistingBooksByHash(it, existingSeries) }
+      .filter {
+        if (it.second.size > 1) logger.debug { "Multiple matches: ${it.second}, skipping entry" }
+        it.second.size == 1
+      }
+      .map { it.first to it.second.first() }
+      .map { it.first.copy(id = it.second.id, seriesId = existingSeries.id, deleted = false) }.toList()
+
+    if (booksToRestore.isNotEmpty()) {
+      logger.info { "Restoring deleted books: $booksToRestore" }
+      bookRepository.updateMany(booksToRestore)
+      updateBookMetadataTitle(booksToRestore)
+    }
+
+    return booksToRestore
+  }
+
+  private fun updateExistingBook(newBooks: Collection<Book>, existingBooks: Collection<Book>) {
+    newBooks.forEach { newBook ->
+      logger.debug { "Trying to match scanned book by url: $newBook" }
+      existingBooks.find { it.url == newBook.url }?.let { existingBook ->
+        logger.debug { "Matched existing book: $existingBook" }
+
+        if (isModified(newBook.fileLastModified, existingBook.fileLastModified)) {
+          logger.info { "Book changed on disk, update and reset media status: $existingBook" }
+          val updatedBook = existingBook.copy(
+            fileLastModified = newBook.fileLastModified,
+            fileSize = newBook.fileSize,
+            fileHash = newBook.fileHash
+          )
+          mediaRepository.findById(existingBook.id).let {
+            mediaRepository.update(it.copy(status = Media.Status.OUTDATED))
           }
+          bookRepository.update(updatedBook)
         }
       }
-    }.also { logger.info { "Library updated in $it" } }
+    }
+  }
+
+  private fun deleteNoLongerExistingBooks(newBooks: Collection<Book>, existingBooks: Collection<Book>) {
+    val newBooksUrls = newBooks.map { it.url }
+    existingBooks.filterNot { existingBook -> newBooksUrls.contains(existingBook.url) }
+      .let { books ->
+        logger.info { "Deleting books not on disk anymore: $books" }
+        bookRepository.softDeleteByBookIds(books.map { it.id })
+      }
+  }
+
+  private fun matchExistingBooksByHash(bookToCheck: Book, existingSeries: Series): Collection<Book> {
+    var existingBooks = bookRepository.findByFileHashAndSizeIncludeDeleted(bookToCheck.fileHash, bookToCheck.fileSize)
+      // check if book exists in case it was deleted but library analysis wasn't done yet
+      .filter { it.deleted || Files.notExists(it.path()) }
+    if (existingBooks.size > 1) {
+      logger.debug { "Multiple matches: $existingBooks, limiting matching to library of the series" }
```

Comment:
we should perform the filter at the repository find in this case, instead of filtering afterwards